FullScreen Unfullscreen button added with resizing functionality

ksraj123 opened the pull request:
Full screen / Unfullscreen button added with resizing functionality.

Fixes #92 and https://github.com/llaske/sugarizer/issues/602

- [x] The fullscreen button should be on the right of the toolbar, near the Stop button.
- [x] The fullscreen button should be always visible, including in small screen size. Could require to adapt media queries in CSS.
- [x] The size of the board should be increase in fullscreen mode. So it means that the resize event should be call directly or indirectly when clicking on fullscreen/unfullscreen button.
- [x] The unfullscreen button should be visible only when the screen is in fullscreen mode.
- [x] The unfullscreen button should be attach to the right of the screen. It's better if it's not at the same location than the Stop button to avoid side effect on multiple click.

on small **screen size 500 x 300**
![smallscreen](https://user-images.githubusercontent.com/39027928/75077191-0a6a1a80-5528-11ea-82ca-1821ba04aec9.gif)


**resizing in action** - size of board increases in fullscreen mode

![resizing](https://user-images.githubusercontent.com/39027928/75081832-bd417500-5536-11ea-9472-1b26cb0fc6b2.gif)

The dark flashes are due to the screen recording software. Sorry for that.

**desktop size screen**
![largescreen](https://user-images.githubusercontent.com/39027928/75077197-105ffb80-5528-11ea-91be-991ae1f04f3d.gif)

ksraj123 commented:
**Review of Changes Made**

Total 7 files changed.
Below is the description of changes made to each file.

**translations.js** - Strings added which will be displayed in the tooltips of fullscreen and unfullscreen button.

**Navbar.js** - FullScreen/UnFullScreen Button Added to the Navbar Component. Conditional logic added to render either the navbar or  the unfullscreen button depending upon if it is in fullscreen mode right now which is passed in as a prop from Sugarizer component and is a part of its state.

**Navbar.css** - styles added for fullscreen and unfullscreen button, the unfullscreen button is made to align over the fullscreen button and right to the stop button. The fullscreen button stays visible in the toolbar even if width of screen is less than ideal.

**fullscreen.svg** / **unfullscreen.svg** - two new files added in the icons directory, icons which were already being used in the qr code activity are reused.

llaske commented:
@thingthing could you review this fix and compare to https://github.com/llaske/ExerciserReact/pull/93?

thingthing commented on src/components/Navbar.js:
```diff
@@ -55,71 +55,84 @@ class Navbar extends Component {
-				!this.props.location.pathname.startsWith('/play') &&
-				!this.props.location.pathname.startsWith('/scores') &&
+			!this.props.inFullscreenMode? 
+				(<div id="main-toolbar" className="toolbar">
```
Since it starting to be a little big in this component, i would maybe create a new component: Toolbar (or MainToolbar) with the toolbar in it, it would make this code more readable.

ksraj123 replied:
Moved it to a new component `MainToolbar`

thingthing commented on src/containers/Router.js:
```diff
@@ -24,52 +24,79 @@ import '../css/index.css';
+			this.resizeBoard(mainContainer.offsetHeight);
+		} else {
+			if (this.state.zoom !== 100)	this.setState({zoom: 100});
+		}
```
It's a nitpick, but you can just use a else if syntax here:
```js
} else if (this.state.zoom !== 100) {
   this.setState({zoom: 100});
}
```

ksraj123 replied:
Fixed

thingthing commented on src/containers/Router.js:
```diff
@@ -24,52 +24,79 @@ import '../css/index.css';
+
+	render(){
+		const props = this.props;
+		const { onUpdate, onSharedResult, inEditMode } = props;
```
You can just do
```js
const { onUpdate, onSharedResult, inEditMode } = this.props;
```
Since the `{...props}` in the Route should be the props from the render function, not this one

ksraj123 replied:
Issue fixed.

thingthing commented on src/containers/Router.js:
```diff
@@ -24,52 +24,79 @@ import '../css/index.css';
-			</Switch>
-		</div>
-	)
+class Main extends Component {
```
I think you can just make this in a functional component: the `componentDidUpdate` is would be the same here as juste making your computation in the functionnal component and it would be the same with your state.
If you want to be sure to note compute it too many times, you can set it in a `useMemo`.
Here you rerender your component two times everytime (since `componentDidUpdate` is called after the render, and `setState` create a rerender).

ksraj123 replied:
The issue is that I need to get the height of the component to perform the computation. I think the height could be found once the component is already rendered then it will be rendered again with the calculated value of zoom passed in. So I guess rendering twice cant be avoided. 

If making it a functional component is wanted then I guess I could maintain a height property in parent state, add componentDidUpdate in parent to find the height of child and setState then pass it to the child functional component as a prop and calculate zoom value based on it in the child component. But still it will be rendered twice.

Also I think finding height of stateless component is a little trick because I think functional components do not support refs. But I guess I could wrap it in a wrapper div inside render and attach the ref to wrapper.

Requesting suggestion.

thingthing replied:
You can use this for example: https://github.com/Swizec/useDimensions

ksraj123 replied:
Suggested change Made.

llaske commented:
Note that the Fullscreen button should be between the Stop button and the Help. 
The Stop button must be to the right of the screen.

ksraj123 commented:
Changes made
- Changed Position of the Fullscreen button, 
- Refactored code and made adjustments according to suggestions.
- Added the Fullscreen button to tutorial.

![exercizerTutorial](https://user-images.githubusercontent.com/39027928/75352943-8d94c300-58d0-11ea-992d-8a64ea3b42bc.gif)

ksraj123 commented:
#95 fixed no space below the submit button issue only for `GroupAssignmentForm` because this issue was noticeable in only this container on desktop. However, it was apparent in other containers as well on mobile size. It has now been fixed in all containers.

![together](https://user-images.githubusercontent.com/39027928/75364634-08b2a500-58e2-11ea-9b46-eb2cc27fef4d.png)

ksraj123 commented:
Improved zooming functionality

Before
![mcq form before](https://user-images.githubusercontent.com/39027928/75365836-ca1dea00-58e3-11ea-86a8-e6ac0d0732b1.gif)

Now
![mcq form now](https://user-images.githubusercontent.com/39027928/75365847-d013cb00-58e3-11ea-97e0-9b2fc7520dc8.gif)

llaske commented:
It don't work on my side.
In Sugarizer, I've got the following issue in the browser console:
```

react-dom.production.min.js:187 TypeError: Object(...) is not a function
    at n (index.js:20)
    at v (Router.js:29)
    at ye (react-dom.production.min.js:172)
    at je (react-dom.production.min.js:201)
    at Le (react-dom.production.min.js:202)
    at ct (react-dom.production.min.js:211)
    at at (react-dom.production.min.js:210)
    at ot (react-dom.production.min.js:210)
    at tt (react-dom.production.min.js:208)
    at _e (react-dom.production.min.js:206)
ke @ react-dom.production.min.js:187
Se.e.callback @ react-dom.production.min.js:194
ie @ react-dom.production.min.js:143
ee @ react-dom.production.min.js:144
Bt @ react-dom.production.min.js:219
ct @ react-dom.production.min.js:211
at @ react-dom.production.min.js:210
ot @ react-dom.production.min.js:210
tt @ react-dom.production.min.js:208
_e @ react-dom.production.min.js:206
ht @ react-dom.production.min.js:222
dt @ react-dom.production.min.js:223
pt.render @ react-dom.production.min.js:230
(anonymous) @ react-dom.production.min.js:234
It @ react-dom.production.min.js:221
Rt @ react-dom.production.min.js:234
render @ react-dom.production.min.js:236
(anonymous) @ index.js:10
i @ bootstrap 92665c002444647e85b4:19
(anonymous) @ main.d23fae36.js:16270
i @ bootstrap 92665c002444647e85b4:19
(anonymous) @ bootstrap 92665c002444647e85b4:62
(anonymous) @ bootstrap 92665c002444647e85b4:62
index.js:20 Uncaught TypeError: Object(...) is not a function
    at n (index.js:20)
    at v (Router.js:29)
    at ye (react-dom.production.min.js:172)
    at je (react-dom.production.min.js:201)
    at Le (react-dom.production.min.js:202)
    at ct (react-dom.production.min.js:211)
    at at (react-dom.production.min.js:210)
    at ot (react-dom.production.min.js:210)
    at tt (react-dom.production.min.js:208)
    at _e (react-dom.production.min.js:206)
n @ index.js:20
v @ Router.js:29
ye @ react-dom.production.min.js:172
je @ react-dom.production.min.js:201
Le @ react-dom.production.min.js:202
ct @ react-dom.production.min.js:211
at @ react-dom.production.min.js:210
ot @ react-dom.production.min.js:210
tt @ react-dom.production.min.js:208
_e @ react-dom.production.min.js:206
ht @ react-dom.production.min.js:222
dt @ react-dom.production.min.js:223
pt.render @ react-dom.production.min.js:230
(anonymous) @ react-dom.production.min.js:234
It @ react-dom.production.min.js:221
Rt @ react-dom.production.min.js:234
render @ react-dom.production.min.js:236
(anonymous) @ index.js:10
i @ bootstrap 92665c002444647e85b4:19
(anonymous) @ main.d23fae36.js:16270
i @ bootstrap 92665c002444647e85b4:19
(anonymous) @ bootstrap 92665c002444647e85b4:62
(anonymous) @ bootstrap 92665c002444647e85b4:62

```
In Debug mode I've got the error:

![image](https://user-images.githubusercontent.com/981841/75484071-57c30d80-59a8-11ea-9472-e37f1f0fcd27.png)

ksraj123 commented:
useDimensions package was added to determine the height of component which uses hooks and is a hook itself but hooks were introduced in react v16.8.0 while this repo uses v16.4.0. I believe this is what is causing this issue. I think there are no breaking changes between v16.4.0 and v16.8.0 as per the react change logs (https://github.com/facebook/react/blob/master/CHANGELOG.md) so updating the version of react and react-dom in package.json should resolve the issue. I could reproduce this issue in a fresh environment and it got resolved this way.

ksraj123 commented:
@llaske please review

llaske commented:
I can run it now.
BTW your fix cause an issue in drag&drop of items.
See for example in Roman numeral sample. Here's your fix:

![2020-02-28 17-47-25 2020-02-28 17_48_25](https://user-images.githubusercontent.com/981841/75568086-edb37280-5a52-11ea-98ff-c69b50892354.gif)

Here's the dev.sugarizer.org version:

![2020-02-28 17-49-56 2020-02-28 17_51_38](https://user-images.githubusercontent.com/981841/75568124-02900600-5a53-11ea-8267-340d30c55bf9.gif)

ksraj123 commented:
@llaske Please review

Here is a list of changes made -
- Issue with dnd in reorder list fixed.

![dnd fixed](https://user-images.githubusercontent.com/39027928/75637988-e311fc80-5c4f-11ea-86ac-c0f732fab07b.gif)

- Changes made to names of a few classes and ids to better follow react conventions.
- Made more improvements to the zooming functionality.
![zooming improved](https://user-images.githubusercontent.com/39027928/75639149-76016580-5c55-11ea-9349-aacaf383aa57.gif)

- added logic to make changes to each component's styling in fullscreen mode to better utilitze the extra space.
![reorder list comp](https://user-images.githubusercontent.com/39027928/75638663-f4103d00-5c52-11ea-9f10-b3cdc7fa97b9.png)

![matching comp](https://user-images.githubusercontent.com/39027928/75639334-57e83500-5c56-11ea-9ecc-162989e18e0c.png)

- Added margin to top and bottom of a few components to make styling across the application more consistent.

The reason for the issue with drag n drop was because `transform scale` scales only the displayed output and does not modify the box sizing of the element due to which position of element was getting messed up when it was being dragged while `zoom` does exactly what was needed. css `zoom` property is supported by all prominent browsers except firefox. However logic is added for the component's styling to change to make efficient use of the extra space in fullscreen mode. So I think the fullscreen mode serves its purpose on firefox as well.

Here is a walkthrough on firefox -
![firefox walktrough](https://user-images.githubusercontent.com/39027928/75639956-1c029f00-5c59-11ea-9b0c-33fb059d54af.gif)

thingthing commented on src/containers/Builders/FreeTextInputForm.js:
```diff
@@ -481,11 +481,22 @@ class FreeTextInputForm extends Component {
 			answer_error = <span style={{ color: "red" }}><FormattedMessage id={ANSWER_ERROR} /></span>;
 		}
 
+		let fullScreenStyles = {
```
You should create a cs class for this, so you can use it everywhere

ksraj123 replied:
Refactored code according to suggestion.

thingthing commented on src/containers/Players/GroupAssignmentPlayer.js:
```diff
@@ -343,10 +343,10 @@ class GroupAssignmentPlayer extends Component {
 
 		return (
-			<div className="container group-container">
+			<div className="container" id={(this.props.inFullscreenMode? "fullScreenMargin" : "group-container")} >
```
Be careful, you can't have multiple elements with the same id in a html page. If you have the same value, you should set a class instead of an id

ksraj123 replied:
The application could be either in fullscreen mode or not, depending upon that the appropriate id will be assigned. The id for unfullscreen mode `group-container` in this case  is used only here in the entire application. The components to which changes are made are such that only one on them will be displayed at any time so the id in case of fullscreen mode will also appear only once in one page. So I am not sure about how the id is getting repeated in the same page here. However necessary refactoring done to avoid any chance of this happening.

thingthing commented on src/containers/ExerciseList.js:
```diff
@@ -168,12 +168,22 @@ class ExerciseList extends Component {
 			)
 		}
 
+		let fullScreenStyles = {
```
You can use the class here also

ksraj123 replied:
Suggested changes made.

llaske commented:
Thanks the issue is fixed.

BTW I've found another issue.
When the fullscreen button is clicked then unfullscreen button is clicked, the activity palette, the presence palette and the stop button no longer work.

![Peek 03-03-2020 22-26](https://user-images.githubusercontent.com/981841/75821253-372af700-5d9e-11ea-88c4-f16aebd6816d.gif)

ksraj123 commented:
@llaske ?? Requesting review

ksraj123 commented:
@llaske Is there anything more expected from this PR? Requesting Review.

ksraj123 commented:
@llaske Is there anything more expected from this PR? Requesting Review.

llaske commented:
Fixed in https://github.com/llaske/ExerciserReact/pull/93
